File: source/javascript/arena_allocator.js
// Simple bump allocator from an ArrayBuffer
const ARENA_SIZE = 1024;
const arenaBuffer = new ArrayBuffer(ARENA_SIZE);
const arenaView = new DataView(arenaBuffer);
let offset = 0;

function arenaAlloc(size) {
    if (offset + size > ARENA_SIZE) {
        console.log("arena out of memory!");
        return null;
    }
    const ptr = offset;
    offset += size;
    return ptr;
}

function arenaReset() { offset = 0; }

function writeI32(ptr, val) { arenaView.setInt32(ptr, val, true); }
function readI32(ptr) { return arenaView.getInt32(ptr, true); }

const a = arenaAlloc(4);
const b = arenaAlloc(4);
const c = arenaAlloc(4);
const d = arenaAlloc(4);
const e = arenaAlloc(4);

writeI32(a, 100); writeI32(b, 200); writeI32(c, 300); writeI32(d, 400); writeI32(e, 500);

console.log(`slot 0: ${readI32(a)}`);
console.log(`slot 1: ${readI32(b)}`);
console.log(`slot 2: ${readI32(c)}`);
console.log(`slot 3: ${readI32(d)}`);
console.log(`slot 4: ${readI32(e)}`);
console.log(`arena used: ${offset} bytes`);
console.log("arena freed");

File: source/go/arena_allocator.go
package main

import "fmt"

// Simple bump allocator from a large backing byte slice
const arenaSize = 1024

var arena = make([]byte, arenaSize)
var offset = 0

func arenaAlloc(size int) []byte {
	if offset+size > arenaSize {
		fmt.Println("arena out of memory!")
		return nil
	}
	p := arena[offset : offset+size]
	offset += size
	return p
}

func arenaReset() {
	offset = 0
}

// Write/read i32 into a byte slice
func writeI32(buf []byte, val int32) {
	buf[0] = byte(val)
	buf[1] = byte(val >> 8)
	buf[2] = byte(val >> 16)
	buf[3] = byte(val >> 24)
}

func readI32(buf []byte) int32 {
	return int32(buf[0]) | int32(buf[1])<<8 | int32(buf[2])<<16 | int32(buf[3])<<24
}

func main() {
	a := arenaAlloc(4)
	b := arenaAlloc(4)
	c := arenaAlloc(4)
	d := arenaAlloc(4)
	e := arenaAlloc(4)

	writeI32(a, 100)
	writeI32(b, 200)
	writeI32(c, 300)
	writeI32(d, 400)
	writeI32(e, 500)

	fmt.Printf("slot 0: %d\n", readI32(a))
	fmt.Printf("slot 1: %d\n", readI32(b))
	fmt.Printf("slot 2: %d\n", readI32(c))
	fmt.Printf("slot 3: %d\n", readI32(d))
	fmt.Printf("slot 4: %d\n", readI32(e))

	fmt.Printf("arena used: %d bytes\n", offset)

	// Reset the whole arena at once (equivalent to free)
	arenaReset()
	fmt.Println("arena freed")
}
